File: C++/Anagrams.cpp
#include<iostream>
#include<vector>
#include<map>
#include<string>
#include<string.h>
#include<math.h>
#include<algorithm>
using namespace std;
class Solution {
public:
    vector<string> anagrams(vector<string> &strs) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
        vector<string> ret;
        map<string, vector<string>> vm;
		for(int i = 0; i < strs.size(); i++)
		{
			string cur = strs[i];
			sort(cur.begin(), cur.end());
			map<string, vector<string>>::iterator iter = vm.find(cur);
			if(iter != vm.end())
			{
				iter->second.push_back(strs[i]);
			}
			else
			{
				vector<string> list;
				list.push_back(strs[i]);
				vm[cur] = list;
			}	
		}
		
		for(map<string, vector<string>>::iterator iter = vm.begin(); iter != vm.end(); iter++)
		{
			vector<string> list = iter->second;
			if(list.size() <= 1)
				continue;
			for(int i = 0; i < list.size(); i++)
			{
				ret.push_back(list[i]);
			}
		}
		return ret;
    }
};

File: JAVA/Anagrams.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.LinkedList;

class Solution {

	public ArrayList<String> anagrams(String[] strs) {
		// Start typing your Java solution below
		// DO NOT write main() function
		ArrayList<String> ret = new ArrayList<String>();
		if (strs == null || strs.length == 0) {
			return ret;
		}
		HashMap<String, LinkedList<String>> listMap = new HashMap<String, LinkedList<String>>();
		for (String str : strs) {
			char[] chs = str.toCharArray();
			Arrays.sort(chs);
			String ordered = new String(chs);
			if (listMap.containsKey(ordered)) {
				listMap.get(ordered).add(str);
			} else {
				LinkedList<String> list = new LinkedList<String>();
				list.add(str);
				listMap.put(ordered, list);
			}
		}
		for (LinkedList<String> list : listMap.values()) {
			if (list.size() > 1) {
				ret.addAll(list);
			}
		}
		return ret;
	}
}

public class Anagrams {

	/**
	 * @param args
	 */
	public static void main(String[] args) {
		// TODO Auto-generated method stub
		String[] strs = {"and","dan"};
		Solution sol = new Solution();
		System.out.println(sol.anagrams(strs));
	}

}
